sdk/azcore/policy_token_refresh.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azcore
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"sync"
+	"time"
+)
+
+const (
+	bearerTokenPrefix = "Bearer "
+)
+
+// DO NOT USE. This is for internal SDK usage. The TokenRefreshProcessor is used to provide
+// custom implementations for the bearer token policy.
+type TokenRefreshProcessor interface {
+	// IsZeroOrExpired returns a bool if there is a tenant that needs to be initialized, it also
+	// returns a list of tenants to get a token for
+	IsZeroOrExpired(map[string]time.Time) (bool, []string)
+	// ShouldRefresh returns a bool is a token needs to be refreshed and the list of tenants that
+	// need to be refreshed
+	ShouldRefresh(map[string]time.Time) (bool, []string)
+	// GetToken performs the GetToken call and returns the header needed for authentication
+	GetToken(ctx context.Context, cred TokenCredential, tenants []string, opts TokenRequestOptions) (string, error)
+	// Header returns the key for the header in the request
+	Header() string
+}
+
+type tokenRefreshPolicy struct {
+	// cond is used to synchronize token refresh.  the locker
+	// must be locked when updating the following shared state.
+	cond *sync.Cond
+
+	// renewing indicates that the token is in the process of being refreshed
+	renewing bool
+
+	// header contains the authorization header value
+	header string
+
+	// expiresOn is when the token will expire
+	expiresOn map[string]time.Time
+
+	// the following fields are read-only
+	creds       TokenCredential
+	options     TokenRequestOptions
+	implementer TokenRefreshProcessor
+}
+
+// NewTokenRefreshPolicy instantiates a thread safe token refresh policy.
+// Pass in the credential to use, along with a TokenProcessor that will be
+// used to provide a custom implementation for checking the expires_on
+// time of the token, when tokens should be refreshed, and setting custom headers.
+// Additionally specify options for the token to be retreived.
+func NewTokenRefreshPolicy(cred TokenCredential, p TokenRefreshProcessor, opts AuthenticationPolicyOptions) Policy {
+	return &tokenRefreshPolicy{
+		cond:        sync.NewCond(&sync.Mutex{}),
+		creds:       cred,
+		options:     opts.Options,
+		implementer: p,
+	}
+}
+
+func (b *tokenRefreshPolicy) Do(req *Request) (*Response, error) {
+	if req.URL.Scheme != "https" {
+		// HTTPS must be used, otherwise the tokens are at the risk of being exposed
+		return nil, errors.New("token credentials require a URL using the HTTPS protocol scheme")
+	}
+	// create a "refresh window" before the token's real expiration date.
```

Comment:
I think this comment no longer applies to this generic implementation?

Reply:
Removed.